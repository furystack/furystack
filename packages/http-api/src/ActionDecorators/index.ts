export * from "./AllowCors"
export * from "./Authenticate"
export * from "./Authorize"
export * from "./Method"