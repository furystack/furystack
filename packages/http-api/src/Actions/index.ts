export * from "./GetCurrentUser";
export * from "./Login";
export * from "./Logout";
