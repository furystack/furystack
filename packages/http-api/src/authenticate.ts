import '@furystack/logging/dist/InjectorExtension'
import { IncomingMessage } from 'http'
import { Injector } from '@furystack/inject'
import { sleepAsync } from '@sensenet/client-utils'
import { JsonResult, RequestAction } from './Models'
import { HttpUserContext } from './HttpUserContext'

export const Authenticate = () => (action: RequestAction) => {
  return async (i: Injector) => {
    const userContext = i.getInstance(HttpUserContext)
    const authenticated = await userContext.isAuthenticated()
    if (!authenticated) {
      const { url } = i.getInstance(IncomingMessage)
      await sleepAsync(Math.random() * 1000)
      i.logger.warning({
        scope: '@furystack/http-api/@Authenticate()',
        message: `An unauthenticated user has been tried to access to action '${url}' without authentication.`,
        data: {
          url,
        },
      })
      return JsonResult(
        { error: 'unauthorized' },
        401,
        userContext.authentication.enableBasicAuth ? { 'WWW-Authenticate': 'Basic' } : {},
      )
    }
    return await action(i)
  }
}
