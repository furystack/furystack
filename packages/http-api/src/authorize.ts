import '@furystack/logging/dist/InjectorExtension'
import { IncomingMessage } from 'http'
import { Injector } from '@furystack/inject'
import { sleepAsync } from '@sensenet/client-utils'
import { JsonResult, RequestAction } from './Models'
import { HttpUserContext } from './HttpUserContext'

export const Authorize = (...roles: string[]) => (action: RequestAction) => {
  return async (i: Injector) => {
    const userContext = i.getInstance(HttpUserContext)
    const currentUser = await userContext.getCurrentUser()
    const authorized = await userContext.isAuthorized(...roles)
    if (!authorized) {
      const { url } = i.getInstance(IncomingMessage)
      await sleepAsync(Math.random() * 1000)
      i.logger.warning({
        scope: '@furystack/http-api/@Authorize()',
        message: `User '${currentUser.username}' has been tried to access to action '${url}' without the required roles.`,
        data: {
          user: currentUser,
          requiredRoles: roles,
        },
      })
      return JsonResult({ error: 'forbidden' }, 403)
    }
    return await action(i)
  }
}
