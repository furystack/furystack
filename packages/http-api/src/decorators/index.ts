export * from './authenticate'
export * from './authorize'
