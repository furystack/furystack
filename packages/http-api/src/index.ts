export * from "./ActionDecorators";
export * from "./Actions";
export * from "./ErrorAction";
export * from "./IdentityService";
export * from "./HttpApi";
export * from "./MetadataAction";
export * from "./NotFoundAction";
export * from "./Models";
export * from "./RequestAction";
export * from "./RequestContext";
export * from "./RootAction";
export * from "./RouteAction";
