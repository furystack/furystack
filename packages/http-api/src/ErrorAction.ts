import { IContext } from "@furystack/core";
import { IncomingMessage, ServerResponse } from "http";
import { Authorize } from "./ActionDecorators/Authorize";
import { RequestAction } from "./RequestAction";

export class ErrorAction extends RequestAction {
    public exec(incomingMessage: IncomingMessage, serverResponse: ServerResponse, getContext: () => IContext): Promise<void> {
        throw new Error("Method not implemented.");
    }
    public async returnError(incomingMessage: IncomingMessage, serverResponse: ServerResponse, _getContext: () => IContext, error: any): Promise<void> {
        serverResponse.writeHead(500, "Server error",
            { "Content-Type": "application/json" },
        );
        serverResponse.write(JSON.stringify({ message: error.message, url: incomingMessage.url, stack: error.stack }));
        serverResponse.end();
    }
    public segmentName: string = "";

    constructor() {
        super();
    }
}
