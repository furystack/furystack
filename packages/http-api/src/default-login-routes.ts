import { Constructable } from '@furystack/inject'
import { IncomingMessage } from 'http'
import { parse } from 'url'
import { GetCurrentUser } from './Actions/GetCurrentUser'
import { LoginAction } from './Actions/Login'
import { LogoutAction } from './Actions/Logout'
import { IRequestAction } from './Models'

/**
 * Default routes for /login, /logout and /getCurrentUser
 * @param msg The incoming HTTP Message
 */
export const defaultLoginRoutes: (msg: IncomingMessage) => Constructable<IRequestAction> | undefined = msg => {
  const urlPathName = parse(msg.url || '', true).pathname
  switch (urlPathName) {
    case '/currentUser':
      return GetCurrentUser
    case '/login':
      return LoginAction
    case '/logout':
      return LogoutAction
  }
  return undefined
}
