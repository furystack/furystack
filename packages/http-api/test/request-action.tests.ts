import { JsonResult, PlainTextResult, XmlResult, EmptyResult, BypassResult } from '../src'

describe('RequestActions', () => {
  it('Defaults should match the snapshots', () => {
    expect(JsonResult({ foo: 1 })).toMatchSnapshot()
    expect(PlainTextResult('alma')).toMatchSnapshot()
    expect(XmlResult('alma')).toMatchSnapshot()
    expect(EmptyResult()).toMatchSnapshot()
    expect(BypassResult()).toMatchSnapshot()
  })
})
