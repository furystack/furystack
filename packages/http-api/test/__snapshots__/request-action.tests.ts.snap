// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestActions Defaults should match the snapshots 1`] = `
Object {
  "chunk": "{\\"foo\\":1}",
  "headers": Object {
    "Content-Type": "application/json",
  },
  "statusCode": 200,
}
`;

exports[`RequestActions Defaults should match the snapshots 2`] = `
Object {
  "chunk": "alma",
  "headers": Object {
    "Content-Type": "plain/text",
  },
  "statusCode": 200,
}
`;

exports[`RequestActions Defaults should match the snapshots 3`] = `
Object {
  "chunk": "alma",
  "headers": Object {
    "Content-Type": "application/xml;charset=utf-8",
  },
  "statusCode": 200,
}
`;

exports[`RequestActions Defaults should match the snapshots 4`] = `
Object {
  "headers": Object {},
  "statusCode": 200,
}
`;

exports[`RequestActions Defaults should match the snapshots 5`] = `
Object {
  "chunk": "BypassResult",
}
`;
